### PR TITLE
Unstable/cpp cauction gen

### LINKED ISSUE #117: Improve Combinatorial Auction
## Describe the bug
`CombinatorialAuction::generate_instance` still has 200+ lines and many local variables.
It should be refactor to improve readability and reduce the chances for errors.

## Improving direction:
 - [ ] Split `generate_instance` into more functions / classes.
 - [ ] Ideally separate the code sampling data (bids, bundle...) from the code building the SCIP instance.
 - [ ] As much as possible, don't use `std::map`. It is not meant to be used the same way as in Python and is much slower than `std::vector`. Also using `std::vector` as keys can be a bit confusing.
 - [ ] When possible avoid using `while(true){ ... break; }` loops

Overall, there is a lot of of variables like `bundle`, `sub_bundles`, `mask`, `price`, `candidate` which make it particularly difficult to read.
I think we would really gain from defining simple classes to represent these abstractions, with meaningful accessors.

### PR DESCRIPTION
**Pull request by jdumouchelle**
## Pull request checklist

- [x] I have open an issue to discuss the proposed changes. Fix #117
- [ ] I have modified/added tests to cover the new changes/features.
- [ ] I have modified/added the documentation to cover the new changes/features.
- [x] I have ran the tests, checks, and code formatters.

## Proposed implementation

Improved structure and readability of combinatorial auction instance generator in c++.

### CODE REVIEW

**Comment by AntoinePrv** [comment]
Where does this start?

**Comment by jdumouchelle** [comment]
This PR is with respect to `refactor of cpp cauction generator`.  The previous commits `updated cauctions for a subset of the PR comments` was made when we initially closed the PR `Unstable/cpp instance gen`.  So you only need to look at `refactor of cpp cauction generator`.

File: libecole/src/instance/combinatorial-auction.cpp [hunk 1]
```diff
@@ -149,173 +340,85 @@ scip::Model CombinatorialAuctionGenerator::generate_instance(Parameters paramete
...
 			}
-			bids_bundle.push_back(bund_copy);
-			bids_price.push_back(p);
+			bids[bid_index] = std::make_tuple(bund_copy, p);
 			++bid_index;
 		}
-
 	}  // loop to get bids
 
 	// create scip model
```
Comment (AntoinePrv) [inline libecole/src/instance/combinatorial-auction.cpp:406]: Can we split the model creation function into a separate function?

File: libecole/src/instance/combinatorial-auction.cpp [hunk 2]
```diff
@@ -149,173 +340,85 @@ scip::Model CombinatorialAuctionGenerator::generate_instance(Parameters paramete
...
 
-		// add bids
 		for (auto const& [b, p] : bidder_bids) {
 			auto bund_copy = b;
 			if (dummy_item) {
 				bund_copy.push_back(dummy_item);
 			}
-			bids_bundle.push_back(bund_copy);
-			bids_price.push_back(p);
+			bids[bid_index] = std::make_tuple(bund_copy, p);
```
Comment (AntoinePrv) [inline libecole/src/instance/combinatorial-auction.cpp:401]:
```suggestion
			bids[bid_index] = std::tuple{std::move(bund_copy), p};
```

File: libecole/src/instance/combinatorial-auction.cpp [hunk 3]
```diff
@@ -149,173 +340,85 @@ scip::Model CombinatorialAuctionGenerator::generate_instance(Parameters paramete
...
+			parameters.resale_factor,
+			parameters.max_n_sub_bids,
+			parameters.n_bids);
+
+		// get dummy item if required
 		std::size_t dummy_item = 0;
 		if (bidder_bids.size() > 2) {
 			dummy_item = parameters.n_items + n_dummy_items;
 			++n_dummy_items;
 		}
```
Comment (AntoinePrv) [inline libecole/src/instance/combinatorial-auction.cpp:394]: Can this be a function `add_dummy_items`?

File: libecole/src/instance/combinatorial-auction.cpp [hunk 4]
```diff
@@ -149,173 +340,85 @@ scip::Model CombinatorialAuctionGenerator::generate_instance(Parameters paramete
...
 			if (dummy_item) {
 				bund_copy.push_back(dummy_item);
 			}
-			bids_bundle.push_back(bund_copy);
-			bids_price.push_back(p);
+			bids[bid_index] = std::make_tuple(bund_copy, p);
 			++bid_index;
 		}
-
 	}  // loop to get bids
```
Comment (AntoinePrv) [inline libecole/src/instance/combinatorial-auction.cpp:404]: Can this be a function `add_bids`?

**Review by AntoinePrv — COMMENTED** [review]
Thanks @jdumouchelle ?
Do you think we could further split into a few more functions?

File: libecole/src/instance/combinatorial-auction.cpp [hunk 5]
```diff
@@ -122,8 +305,18 @@ template <typename String> void warning(String&& message, bool do_log) {
 
 scip::Model CombinatorialAuctionGenerator::generate_instance(Parameters parameters, RandomEngine& random_engine) {
 
-	assert(parameters.min_value >= 0 && parameters.max_value >= parameters.min_value);
-	assert(parameters.add_item_prob >= 0 && parameters.add_item_prob <= 1);
+	// check that parameters are valid
+	if (!(parameters.min_value >= 0 && parameters.max_value >= parameters.min_value)) {
```
[the remark this answers is not in the document]
Comment (jdumouchelle) [reply libecole/src/instance/combinatorial-auction.cpp:309]: Yes, we actually should do that

**Comment by jdumouchelle** [comment]
> Can we split the model creation function into a separate function?
> Can this be a function `add_dummy_items`?
> Can this be a function `add_bids`?

Yes, I can make all of these into separate functions.

**Pull request merged**